### PR TITLE
Support 8x resolution by hijacking cat in webui

### PR DESCRIPTION
**Pull request by nono-Sang**
存在的问题：在 sd-webui 中，onediff 仅支持 64 倍分辨率的图片，对应的 issue 是 https://github.com/siliconflow/onediff/issues/832 （本质原因是，非 64 倍分辨率时，unet 上下采样所产生的张量维度误差会导致 `cat` 操作失败）。

解决方法：发现 sd-webui 是通过劫持 `torch.cat` 来应对这一问题以支持 8 倍分辨率的，所以对应地，只需要在 onediff 中对 `oneflow.cat` 进行劫持即可。

### CODE REVIEW

**Comment by doombeaker** [comment]
请 @lijunliangTG 解决下 CI 里的那个报错问题。

File: onediff_sd_webui_extensions/onediff_hijack.py [hunk 1]
```diff
@@ -1,3 +1,35 @@
+import oneflow
+import compile_ldm
```
Comment (ccssu) [inline onediff_sd_webui_extensions/onediff_hijack.py:2]: 这个 compile_ldm 的意义是，没找到 `xxx = compile_ldm.flow` or `compile_ldm.flow.cat 调用` 可以介绍下吗

Comment (ccssu) [reply onediff_sd_webui_extensions/onediff_hijack.py:2]: [ldm.modules.diffusionmodules.openaimodel.th = sd_hijack_unet.th](https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/1c0a0c4c26f78c32095ebc7f8af82f5c04fca8c0/modules/sd_hijack.py#L70) 这个应该没有做取消 hijack . 

直接设置  oneflow.cat = OneFlowHijackForUnet.cat   吧，  OneFlowHijackForUnet.cat  兼容 oneflow.cat 就好了，
在  def do_hijack(): 函数中 设置下。

Comment (nono-Sang) [reply onediff_sd_webui_extensions/onediff_hijack.py:2]: 1. 是的，sd-webui 没有取消对 `torch` 的 hijack（也没有定义 `undo_hijack` 操作）。我想的是，既然 onediff 定义了 `undo_hijack` 函数，那是不是需要在其内部取消对 `oneflow` 的 hijack 呢？
2. 我的看法是：因为我们的目的是拦截 `compile_ldm` 和 `compile_sgm` 的 `flow` 包，即 `flow.cat` 使用自定义的 `cat`，其他操作保持不变，即 `OneFlowHijackForUnet` 类的功能，所以是：
```py
hijack_flow = OneFlowHijackForUnet()
compile_ldm.flow = hijack_flow 
compile_sgm.flow = hijack_flow
```
如果直接在 `do_hijack` 函数中设置 `oneflow.cat = OneFlowHijackForUnet().cat`，会出现以下报错：
```
[ERROR](GRAPH:OneflowGraph_0:OneflowGraph) building graph got error.
ERROR [2024-05-06 10:57:08] xxx/onediff/src/onediff/infer_compiler/oneflow/utils.py:23 - Exception in forward: e=RecursionError('maximum recursion depth exceeded while calling a Python object')
```
不知道是不是影响到了其他地方的 `oneflow.cat`（因为这个修改是全局的）。

**Comment by wangerlie** [comment]
因为我这边没有webui的环境，然后安装时一直出现网络问题，就先让 @lijunliangTG 帮忙测试了。
img_v3_02ak_d06e00bd-2139-4e2f-9463-594f06f7860g  

img_v3_02ak_cb4476b0-776c-4f10-8db3-1a7baaec6ebg  
  
生成结果
img_v3_02ak_0dd66617-b42b-4aff-a44e-6ba9799b72eg

File: onediff_sd_webui_extensions/onediff_hijack.py (hunk 1 above)
Comment (marigoold) [reply onediff_sd_webui_extensions/onediff_hijack.py:2]: > 如果直接在 `do_hijack` 函数中设置 `oneflow.cat = OneFlowHijackForUnet().cat`，会出现以下报错：
> 
> ```
> [ERROR](GRAPH:OneflowGraph_0:OneflowGraph) building graph got error.
> ERROR [2024-05-06 10:57:08] xxx/onediff/src/onediff/infer_compiler/oneflow/utils.py:23 - Exception in forward: e=RecursionError('maximum recursion depth exceeded while calling a Python object')
> ```
> 
> 不知道是不是影响到了其他地方的 `oneflow.cat`（因为这个修改是全局的）。

我猜这里是 OneFlowHijackForUnet.cat 内部调用了 oneflow.cat，然后在调用 oneflow.cat 的时候又回到了 OneFlowHijackForUnet.cat 导致死循环

**Review by marigoold — APPROVED** [review]

**Comment by lijunliangTG** [comment]
image
Picture after onediff acceleration
image
Picture of the original seed
image
shell output
image

File: onediff_sd_webui_extensions/onediff_hijack.py (hunk 1 above)
Comment (nono-Sang) [reply onediff_sd_webui_extensions/onediff_hijack.py:2]: 👍 应该是这个原因，递归爆栈了

**Review by ccssu — APPROVED** [review]

**Review by doombeaker — APPROVED** [review]

File: onediff_sd_webui_extensions/onediff_hijack.py [hunk 2]
```diff
@@ -1,3 +1,35 @@
...
+        if hasattr(oneflow, item):
+            return getattr(oneflow, item)
+
+        raise AttributeError(f"'{type(self).__name__}' object has no attribute '{item}'")
+
+    def cat(self, tensors, *args, **kwargs):
+        if len(tensors) == 2:
+            a, b = tensors
+            if a.shape[-2:] != b.shape[-2:]:
+                a = oneflow.nn.functional.interpolate(a, b.shape[-2:], mode="nearest")
```
Comment (strint) [inline onediff_sd_webui_extensions/onediff_hijack.py:25]: 写在 python 层，会导致 graph trace 固定这里的计算逻辑。切换分辨率时可能有 bug
